### PR TITLE
Make LiveTestWidgetsFlutterBinding work with setSurfaceSize and live tests

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR fixes several bugs related to live tests, adds more tests, and completes the documentation of several methods related to pointer events. More specifically:

The several tests for `LiveTestWidgetsFlutterBinding` added by #83337 were discovered by https://github.com/flutter/flutter/pull/82712#issuecomment-871534510 to fail when `setSurfaceSize` is called. This PR fixed these problems and added a `setSurfaceSize` version for these tests.

This PR fixes few problems of live tests:

  * https://github.com/flutter/flutter/issues/82542 reported that simulated pointer events do not work on live tests (for example, some code lab tests are broken).
  * Tapping the screen on live tests should print the hit widgets, instead the events are currently forwarded to the app and painted.
  * The painted event positions currently do not match the positions where the user interact.

This PR changed the testing widget used in `animation/live_binding_test.dart`. The target widget is now an `InkWell` to clearly ensure that the `hitTest` occurs at the intended local position.

Finally, this PR updated some documentations for some methods, most importantly, whether a pointer event should be in the global coordinate system or the local one.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/86449

**Comment by skia-gold** [comment]
Gold has detected about 6 new digest(s) on patchset 2.
View them at https://flutter-gold.skia.org/cl/github/86449

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/86449

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/86449

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 5.
View them at https://flutter-gold.skia.org/cl/github/86449

**Comment by skia-gold** [comment]
Gold has detected about 3 new digest(s) on patchset 7.
View them at https://flutter-gold.skia.org/cl/github/86449

File: packages/flutter_test/lib/src/binding.dart [hunk 1]
```diff
@@ -462,15 +462,38 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
...
   /// Dispatch an event to the targets found by a hit test on its position,
   /// and remember its source as [pointerEventSource].
   ///
-  /// This method sets [pointerEventSource] to `source`, runs
+  /// This method sets [pointerEventSource] to `source`, forwards the call to
   /// [handlePointerEvent], then resets [pointerEventSource] to the previous
   /// value.
+  ///
+  /// If `source` is [TestBindingEventSource.device], then the `event`
+  /// is based in the global coordinate system and the event is likely
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart:479]: Maybe clarify what the global and what the local (local to what?) coordinate system is?

Comment (goderbauer) [reply packages/flutter_test/lib/src/binding.dart:479]: Similar to how it is documented on the localToGlobal/globalToLocal methods.

File: packages/flutter_test/lib/src/binding.dart [hunk 2]
```diff
@@ -482,7 +505,7 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
   /// to the previous value.
   @protected
   void withPointerEventSource(TestBindingEventSource source, VoidCallback task) {
-    final TestBindingEventSource previousSource = source;
+    final TestBindingEventSource previousSource = _pointerEventSource;
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart:508]: wow, good catch!

File: packages/flutter_test/lib/src/binding.dart [hunk 3]
```diff
@@ -1523,8 +1550,9 @@ class LiveTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
         break;
       case TestBindingEventSource.device:
         if (deviceEventDispatcher != null) {
+          final PointerEvent localEvent = event.copyWith(position: globalToLocal(event.position));
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart:1553]: This line probably deserves a comment explaining why the position needs to be transformed.

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 4]
```diff
@@ -59,6 +59,17 @@ export 'package:test_api/test_api.dart' hide
 /// Signature for callback to [testWidgets] and [benchmarkWidgets].
 typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
 
+// [Iterable.lastWhere] but returns null if not found.
+T? _lastWhereOrNull<T>(Iterable<T> list, bool Function(T) condition) {
+  try {
+    return list.lastWhere(condition);
+  } catch(e) {
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:66]: nit: space before (

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 5]
```diff
@@ -59,6 +59,17 @@ export 'package:test_api/test_api.dart' hide
 /// Signature for callback to [testWidgets] and [benchmarkWidgets].
 typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
 
+// [Iterable.lastWhere] but returns null if not found.
+T? _lastWhereOrNull<T>(Iterable<T> list, bool Function(T) condition) {
+  try {
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:64]: This is using try/catch for actual control flow, let's not do that :)

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 6]
```diff
@@ -800,15 +811,15 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
...
-        binding.renderViewElement!,
-        skipOffstage: true,
-      ).cast<Element?>().lastWhere(
-        (Element? element) => element!.renderObject == innerTarget,
-        orElse: () => null,
+      final Element? innerTargetElement = _lastWhereOrNull(
+        collectAllElementsFrom(
+          binding.renderViewElement!,
+          skipOffstage: true,
+        ).cast<Element>(),
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:818]: Why this cast? `collectAllElementsFrom` already returns an `Iterable<Element>`, I think.

File: packages/flutter_test/lib/src/widget_tester.dart [hunk 7]
```diff
@@ -800,15 +811,15 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
         .map((HitTestEntry candidate) => candidate.target)
         .whereType<RenderObject>()
         .first;
-      final Element? innerTargetElement = collectAllElementsFrom(
-        binding.renderViewElement!,
-        skipOffstage: true,
-      ).cast<Element?>().lastWhere(
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:806]: What was wrong with casting this to `Element?` and then using the real `lastWhere` with on `orElse: () => null`?

**Review by goderbauer — COMMENTED** [review]
Thanks so much for debugging this!

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 5 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/widget_tester.dart:64]: I would, but somehow normal cast with `lastWhere` throws strange null assertions. I guess I'll write a manual traversal then.

File: packages/flutter_test/lib/src/widget_tester.dart (hunk 7 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/widget_tester.dart:806]: It somehow throws strange null assertions, saying `null can not be cast to Element`. I guess I'll write a manual traversal then.

File: packages/flutter_test/lib/src/binding.dart (hunk 3 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/binding.dart:1553]: I added 
```dart
          // The pointer events received with this source has a global position
          // (see [handlePointerEventForSource]). Transform it to the local
          // coordinate space used by the testing widgets.
```

File: packages/flutter_test/lib/src/binding.dart (hunk 1 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/binding.dart]: I documented it in `TestWidgetsFlutterBinding` and let related methods refer to it.

File: packages/flutter_test/lib/src/binding.dart [hunk 8]
```diff
@@ -119,6 +119,20 @@ mixin TestDefaultBinaryMessengerBinding on BindingBase, ServicesBinding {
...
+/// sizes, while the testing widget is still told the same size to ensure
+/// consistent results. Consequently, code that deals with positions (such as
+/// pointer events or painting) must distinguish between two coordinate spaces:
+///
+///  * The _local coordinate space_ is the one used by the testing widget
+///    (typically an 800 by 600 window, but can be altered by [setSurfaceSize]).
+///  * The _global coordinate space_ is the one used by the device.
+///
+/// Positions can be transformed between coordinate spaces with [localToGlobal]
+/// and [globalToLocal].
```
Comment (goderbauer) [inline packages/flutter_test/lib/src/binding.dart:135]: Nice explanation! 👍

Comment (goderbauer) [reply packages/flutter_test/lib/src/binding.dart:135]: Maybe add a note that global and local coordinate space are usually identically when run in the headless flutter_tester.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter_test/lib/src/binding.dart (hunk 8 above)
Comment (dkwingsmt) [reply packages/flutter_test/lib/src/binding.dart:135]: Really? I thought they are still 800x600 and 2400x1800 respectively, which still require transforming. Also I worry if that comment means people don't need to worry about transforming in normal cases - they always should!

Comment (goderbauer) [reply packages/flutter_test/lib/src/binding.dart:135]: Ok, let's leave the comment as is.

**Comment by zanderso** [comment]
It looks like some issue with the goldens precipitated by this change is affecting rolls from Engine and Plugins:

See:
https://github.com/flutter/flutter/pull/86687
https://github.com/flutter/flutter/pull/86690

**Comment by Piinks** [comment]
@dkwingsmt please triage images before manually merging PRs. They end up causing errant golden file failures for other PRs.

Is it possible that this change causes an inconsistent image? It look like it has been different for each commit following this change, although it could be caused by the untriaged image errors. They appear to be triaged now, so if they continue to vary for every commit there may be something wrong with the test.

**Pull request merged**